### PR TITLE
shell: devmem: add devmem dump subcommand

### PR DESCRIPTION
**Pull request by cfriedt**
2 commits:
* move `getopt()` declaration to `<zephyr/posix/unistd.h>`
* shell: devmem: add devmem dump subcommand

This allows the caller to dump a region of memory rather than dumping one byte (word, etc) at a time. Additionally, it respects alignment requirements so it works for e.g. 32-bit register accesses.

E.g.
```
uart:~$ devmem dump -a 0x0 -s 42 -w 8
00000000: 18 35 00 20 ed 9a 00 00  f1 89 00 00 85 9a 00 00 |.5. .... ........|
00000010: 85 9a 00 00 85 9a 00 00  85 9a 00 00 00 00 00 00 |........ ........|
00000020: 00 00 00 00 00 00 00 00  00 00                   |........ ..      |
```

### CODE REVIEW

**Comment by cfriedt** [comment]
* move `getopt()` declaration from `<getopt.h>` to `<unistd.h>` (where POSIX specifies)
* include `<unistd.h>` instead of `<getopt.h>`

File: samples/subsys/shell/shell_module/src/main.c [hunk 1]
```diff
@@ -13,6 +13,12 @@
 #include <zephyr/usb/usb_device.h>
 #include <ctype.h>
 
+#ifdef CONFIG_ARCH_POSIX
+#include <unistd.h>
+#else
+#include <zephyr/posix/unistd.h>
+#endif
```
Comment (mbolivar-nordic) [inline samples/subsys/shell/shell_module/src/main.c:20]: Having to do this everywhere is kind of yucky. Can't we add zephyr/posix to the include path if the posix subsystem is enabled or so?

Comment (cfriedt) [reply samples/subsys/shell/shell_module/src/main.c:20]: It sure is!

Part of the LTSv3 POSIX API Roadmap is to eliminate the mutually-exclusiveness of CONFIG_ARCH_POSIX and CONFIG_POSIX_API.

So although this seems like technical debt, it's actually part of the process of fixing several years worth of other technical debt / neglect.

Eventually, to use POSIX APIs, we will need to simply use CONFIG_POSIX_API=y and use the actual standard paths.

**Review by mbolivar-nordic — APPROVED** [review]
This looks reasonable to me and I can't see how it could hurt anything.

**Review by nordic-krch — APPROVED** [review]

File: lib/posix/getopt/getopt.h [hunk 2]
```diff
@@ -58,27 +54,6 @@ void getopt_init(void);
...
- *			the option character.
- * @return		If options have been detected that is not in @p options
- *			function will return '?'.
- *			If function encounters an option with a missing
- *			argument, then the return value depends on the first
- *			character in optstring: if it is ':', then ':' is
- *			returned; otherwise '?' is returned.
- * @return -1		If all options have been parsed.
- */
-int getopt(int nargc, char *const nargv[], const char *ostr);
```
Comment (enjiamai) [inline lib/posix/getopt/getopt.h:80]: How about keeping these descriptions?

**Review by enjiamai — APPROVED** [review]

**Review by jakub-uC — APPROVED** [review]

File: subsys/shell/modules/devmem_service.c [hunk 3]
```diff
@@ -26,6 +32,122 @@ static bool littleendian;
...
+static int memory_dump(const struct shell *sh, mem_addr_t phys_addr, size_t size, uint8_t width)
+{
+	uint32_t value;
+	size_t data_offset;
+	mm_reg_t addr;
+	const size_t vsize = width / BITS_PER_BYTE;
+	uint8_t data[SHELL_HEXDUMP_BYTES_IN_LINE];
+
+#if defined(CONFIG_MMU) || defined(CONFIG_PCIE)
+	device_map((mm_reg_t *)&addr, phys_addr, size, K_MEM_CACHE_NONE);
```
Comment (jhedberg) [inline subsys/shell/modules/devmem_service.c:48]: Shouldn't there be a respective `unmap()` call before returning from this function, or it's fine to leave the memory mapped after you're done with it?

Comment (cfriedt) [reply subsys/shell/modules/devmem_service.c:48]: That's a good question. This part was copied from the existing `devmem read`

**Comment by jakub-uC** [comment]
Does it make sense to move the other declarations as well: `getopt_long` and `getopt_long_only` to `unistd.h`?

**Comment by cfriedt** [comment]
> Does it make sense to move the other declarations as well: `getopt_long` and `getopt_long_only` to `unistd.h`?

Those are non-standard 🥲

**Pull request merged**